src/volatile_memory.rs

```diff
@@ -1261,24 +1270,59 @@ where
             //   a slice and thus has to live outside of guest memory (there can be more slices to
             //   guest memory without violating rust's aliasing rules)
             // - size is always a multiple of alignment, so treating *const T as *const u8 is fine
-                unsafe { copy_slice(destination.as_ptr(), buf.as_ptr() as *const u8, total) };
+                unsafe { copy_to_volatile_slice(&destination, buf.as_ptr() as *const u8, total) };
             self.bitmap.mark_dirty(0, count);
         } else {
-            let mut addr = self.addr;
-            for &v in buf.iter().take(self.len()) {
-                // SAFETY: write_volatile is safe because the pointers are range-checked when
-                // the slices are created, and they never escape the VolatileSlices.
-                // ptr::add is safe because get_array_ref() validated that
-                // size_of::<T>() * self.len() fits in an isize.
-                unsafe {
-                    write_volatile(addr as *mut Packed<T>, Packed::<T>(v));
-                    addr = addr.add(self.element_size());
-                }
-            }
+            // SAFETY: write_volatile is safe because the pointers are range-checked when
+            // the slices are created, and they never escape the VolatileSlices.
+            // ptr::add is safe because get_array_ref() validated that
+            // size_of::<T>() * self.len() fits in an isize.
+            unsafe { self.write_volatile(buf) };
+        }
+    }
+
+    unsafe fn read_volatile_aref(buf: &mut [T], mut addr: *const Packed<T>, count: usize) -> usize {
+        let mut i = 0;
+
+        for v in buf.iter_mut().take(count) {
+            // SAFETY: read_volatile is safe because the pointers are range-checked when
+            // the slices are created, and they never escape the VolatileSlices.
+            // ptr::add is safe because get_array_ref() validated that
+            // size_of::<T>() * self.len() fits in an isize.
+            unsafe {
+                *v = read_volatile(addr).0;
+                addr = addr.add(1);
+            };
+            i += 1;
+        }
+        i
+    }
+
+    unsafe fn write_volatile_aref(&self, buf: &[T], mut addr: *mut Packed<T>, count: usize) {
```

Comment:
why does this include `&self`, but the `read_volatile_aref` does not?

Reply:
I was trying to update the bitmap from within it, changed now to make both of them look same.